flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+
+### `prif_critical_type`
+
+* a derived type provided by the PRIF implementation that is opaque to the
+  compiler and is used for implementing `critical` blocks
+
+# PRIF Procedures
+
+**The PRIF API provides implementations of parallel Fortran features, as specified
+in Fortran 2023. For any given `prif_*` procedure that corresponds to a Fortran
+procedure or statement of similar name, the constraints and semantics associated
+with each argument to the `prif_*` procedure match those of the analogous
+argument to the parallel Fortran feature, except where this document explicitly
+specifies otherwise. For any given `prif_*` procedure that corresponds to a Fortran
+procedure or statement of similar name, the constraints and semantics match those
+of the analogous parallel Fortran feature. In particular, any required synchronization
+is performed by the PRIF implementation unless otherwise specified.**
+
+| **IMPLEMENTATION NOTE**: |
+| ---------------- |
+| The PRIF API is defined as a set of Fortran language procedures and supporting types, and as such an implementation of PRIF cannot be expressed solely in C/C++. However C/C++ can be used to implement portions of the PRIF procedures via calls to BIND(C) procedures. |
+
+Where possible, optional arguments are used for optional parts or different
+forms of statements or procedures. In some cases the different forms or presence
+of certain options change the return type or rank, and in those cases a generic
+interface with different specific procedures is used.
+
+## Common Arguments
+
+There are multiple Common Arguments sections throughout the specification that
+outline details of the arguments that are common for the following sections
+of procedure interfaces.
+
+### Integer and Pointer Arguments
+
+There are several categories of arguments where the PRIF implementation will need
+pointers and/or integers. These fall broadly into the following categories.
+
+1. `integer(c_intptr_t)`: Anything containing a pointer representation where
+   the compiler might be expected to perform pointer arithmetic
+2. `type(c_ptr)` and `type(c_funptr)`: Anything containing a pointer to an
+   object/function where the compiler is expected only to pass it (back) to the
+   PRIF implementation
+3. `integer(c_size_t)`: Anything containing an object size, in units of bytes
+   or elements, i.e. shape, element_size, etc.
+4. `integer(c_ptrdiff_t)`: strides between elements for non-contiguous coarray
+   accesses
+5. `integer(c_int)`: Integer arguments corresponding to image index and
+  stat arguments. It is expected that the most common integer arguments
+  appearing in Fortran code will be of default integer kind, it is expected that
+  this will correspond with that kind, and there is no reason to expect these
+  arguments to have values that would not be representable in this kind.
+6. `integer(c_intmax_t)`: Bounds, cobounds, indices, cosubscripts, and any other
```

Comment:
Quoted from @jeffhammond's response:
> Please do not use c_intmax_t. Just use c_size_t for bounds, etc.

Reply:
**LBL response:** Because integer arguments to many Fortran intrinsics may be any size integer, PRIF needs to support the widest integer the compiler supports. This is by definition c_intmax_t, which POSIX defined for exactly this type of use case. This is often equivalent to c_int64_t, but toolchains are permitted to support even wider types here. Note that in particular, specifying c_size_t here as you suggest would prevent use of 64-bit cobounds and team_numbers on ILP32 platforms (which is still a meaningful concept).

Reply:
> This has not been addressed. intmax_t is bad and should not be used. https://thephd.dev/intmax_t-hell-c++-c has some info and @erichkeane has some strong words about it elsewhere on the internet.

**LBL response:** @jeffhammond Please see our [previous response above](https://github.com/llvm/llvm-project/pull/76088#discussion_r1626612124) from your earlier review. Please provide further discussion on this topic in this thread. 

We acknowledge that there may be implementation-defined types wider than `intmax_t`, but we have no Fortran standard-compliant way to name them. We could be convinced to use `c_int64_t`, if that solves your concern, however we chose `c_intmax_t` to avoid hardcoding a specific limit.

Reply:
ILP32 support is not relevant at this point, but in what context would you want int64_t or larger cobounds on such a system?

Reply:
What's the use case for more than INT_MAX team numbers?